cmd/thanos/rule.go

```diff
@@ -472,49 +490,22 @@ func runRule(
 	}
 
 	// Handle reload and termination interrupts.
-	reload := make(chan struct{}, 1)
 	{
 		cancel := make(chan struct{})
-		reload <- struct{}{} // Initial reload.
-
 		g.Add(func() error {
+			// Initialize rules.
+			if err := reloadRules(logger, ruleFiles, ruleMgr, evalInterval, metrics); err != nil {
+				level.Error(logger).Log("msg", "initialize rules failed", "err", err)
+			}
 			for {
 				select {
-				case <-cancel:
-					return errors.New("canceled")
-				case <-reload:
 				case <-reloadSignal:
-				}
-
-				level.Debug(logger).Log("msg", "configured rule files", "files", strings.Join(ruleFiles, ","))
-				var files []string
-				for _, pat := range ruleFiles {
-					fs, err := filepath.Glob(pat)
-					if err != nil {
-						// The only error can be a bad pattern.
-						level.Error(logger).Log("msg", "retrieving rule files failed. Ignoring file.", "pattern", pat, "err", err)
-						continue
+					if err := reloadRules(logger, ruleFiles, ruleMgr, evalInterval, metrics); err != nil {
+						level.Error(logger).Log("msg", "reload rules by sighup failed", "err", err)
```

Comment:
Not only sighup TBH, can be HTTP `/-/reload` right? So let's fix message here (:

Reply:
@bwplotka Yes. These code hanlde sighup only.

Reply:
Would be nice to treat signal and HTTP reload exactly the same way (as it was before). Why not? We could reuse `reloadSignal` channel.

Reply:
> Would be nice to treat signal and HTTP reload exactly the same way (as it was before). Why not? We could reuse `reloadSignal` channel.

@bwplotka If so, we need another channel to receive the error message for webhandler. So a new struct should wrap reloadSignal and errMsg.  Maybe it is redundant for the current implementation.
BTW, If you persist, I will follow your comment, and any suggestion on the "a new struct"  solution?

Reply:
Add a new struct to handler webhandler. Since we have a reloadSignal as an input paramter, we need select reloadSignal always.
Please help review. Thanks.
@bwplotka

Reply:
LGTM!